## Validation
---

Using another dataset, we can check the accuracy of our model. 

In [ ]:
validationDataURL = "https://raw.githubusercontent.com/woonyee28/mini-project/main/data/validation_data.csv" #Assign link of dataset to variable
validation_tweets = pd.read_csv(validationDataURL)
validation_tweets.head()

id keyword location                                               text  \
0   1     NaN      NaN        deed reason earthquake may allah forgive us   
1   4     NaN      NaN                forest fire near la rong ask canada   
2   5     NaN      NaN  reside ask shelter place notify offic evac she...   
3   6     NaN      NaN        peopl receiv wildfire evac order california   
4   7     NaN      NaN  got sent photo rub alaska smoke wildfire pour ...   

   target  
0       1  
1       1  
2       1  
3       1  
4       1

In [ ]:
validation_tweets.rename(columns={'target': 'isDisaster'},inplace=True) #Rename the column 'label' to 'isDisaster'
validation_tweets.groupby('isDisaster').count() #Group based on the category, and count the number of entries for each category

id  keyword  location  text
isDisaster                               
0           4342     4323      2884  4342
1           3271     3229      2196  3271

In [ ]:
remove_all(validation_tweets, 'text'); #removes all unnecessary characters from [dataset, column]
validation_tweets.drop(['keyword', 'location'], axis=1, inplace=True) #Removes unnecessary rows from the dataset
validation_tweets.head(100) #Verify that all unnecessary characters have been removed from the dataset

id                                               text  isDisaster
0     1        deed reason earthquake may allah forgive us           1
1     4                forest fire near la rong ask canada           1
2     5  reside ask shelter place notify offic evac she...           1
3     6        peopl receiv wildfire evac order california           1
4     7  got sent photo rub alaska smoke wildfire pour ...           1
..  ...                                                ...         ...
95  137  mile backup south acid block right lane exit s...           1
96  138  acid chang life help determin option finance s...           0
97  139  break dead motorcycle car acid happen hagersto...           1
98  141                                  floor marina acid           0
99  143     car even week got fuck car acid ifs fuck drive           1

[100 rows x 3 columns]

In [ ]:
tweets_validation, isDisaster_validation = validation_tweets['text'], validation_tweets['isDisaster']

In [ ]:
isDisaster_validation.head()

0    1
1    1
2    1
3    1
4    1
Name: isDisaster, dtype: int64

In [ ]:
max_len = 75 
trunc_type = "post" 
padding_type = "post" 
oov_token = "<OOV>" 
number_of_tokens = 1000

tokenizer = Tokenizer(num_words = number_of_tokens, lower= 1, oov_token= oov_token) #Initializing Tokenizer
tokenizer.fit_on_texts(tweets_validation) #Process of transforming words into numbers into a dictionary
word_index = tokenizer.word_index

validation_sequences = tokenizer.texts_to_sequences(tweets_validation)

validation_padded = pad_sequences (validation_sequences, maxlen = max_len, padding = padding_type, truncating = trunc_type )

In [ ]:
#accuracy for first model (double confirming)
validation_tweets['first_model_ans'] = pd.DataFrame(np.around(model.predict(validation_padded),decimals=0))
validation_tweets['first_model_acc'] = np.where(validation_tweets['first_model_ans'] == validation_tweets['isDisaster'], 'True', 'False')
count = 0
for i in validation_tweets['first_model_acc']:
  if i=='True':
    count+=1
count/len(validation_tweets['first_model_acc'])

0.49402338105871535

In [ ]:
print(f"Dense architecture loss and accuracy: {model.evaluate(validation_padded,isDisaster_validation)}")
print(f"LSTM architecture loss and accuracy: {model1.evaluate(validation_padded,isDisaster_validation)}")
print(f"Bi-LSTM architecture loss and accuracy: {model2.evaluate(validation_padded,isDisaster_validation)}")

238/238 [==============================] - 0s 1ms/step - loss: 0.9084 - accuracy: 0.4940
Dense architecture loss and accuracy: [0.9083773493766785, 0.4940233826637268]
238/238 [==============================] - 2s 8ms/step - loss: 1.1536 - accuracy: 0.5361
LSTM architecture loss and accuracy: [1.1535794734954834, 0.5360830426216125]
238/238 [==============================] - 3s 11ms/step - loss: 0.7958 - accuracy: 0.5761
Bi-LSTM architecture loss and accuracy: [0.795810341835022, 0.5760917663574219]


As seen from the results, the accuracy using dense network is approximately 49%, while for the LSTM model, it is 53% and lastly for the Bi-LSTM model, it is 57%. 

In this case, the Bi-LSTM architecture had provided a better accuracy. 

However, as our accuracy is only approximately 50%, it shows that while the accuracy is not perfect/good, our test dataset have learn some universal features. 

The decrease in our accuracy of ur results may be due to the difference in the dataset. Furthermore, this may also suggests that our data sample may be unrepresntative which may not have a broad domain. 